kernel/sched.c

```diff
@@ -457,6 +447,25 @@ static void update_cache(int preempt_ok)
 #endif
 }
 
+static bool thread_active_elsewhere(struct k_thread *thread)
+{
+	/* True if the thread is currently running on another CPU.
```

Comment:
I'm surprised threads don't have a "I'm running" flag.

(Then this could just be `thread_running(thread) && thread != _current`)

Reply:
That would be the obvious way to implement this, yeah.  But it requires allocating another flag bit, and we're out of space at the moment.  I'm sensitive to impacting the size of the thread struct for non-SMP use cases.  Really this isn't a hard problem to solve, but on existing devices a quick array loop over 2-4 CPUs is fine for now.

Reply:
Hmm.

Looking at it, we could probably swap out a few of the boolean flags with a state enum instead - most of those flags are one-hot anyway.

But yes, this is fine for now I think.

****

This is _decidedly_ off-topic for this particular PR, but:

> Really this isn't a hard problem to solve, but on existing devices a quick array loop over 2-4 CPUs is fine for now.

My initial testing of Zephyr is showing the opposite, unfortunately. We're fairly bottlenecked on sched_lock in general. (E.g. 2 independent pairs of threads passing 2 semaphores back and forth on a 4-core system end up having _fewer_ total swaps per second than one pair of threads passing one semaphore back and forth.)

(I'm holding off on making a ticket for this until AArch64 SMP support is actually in trunk :-) . Correctness is a rather important prerequisite to performance.)

Reply:
I'd love to look at that report when you're ready.  There's surely stuff we can do to relax contention.  But do at least recognize that k_sem is a VERY thin layer, code that does nothing but pass semaphore tokens back and forth is in some sense doing nothing **but** scheduling, so absent a change in the fundamentals you'd expect the Zephyr design to produce a constant amount of scheduling regardless of CPU count.

Reply:
Yep, it's a rather fundamental design decision that bottlenecks us here unfortunately :-(

In general, the "sea of spinlocks" approach - although better than BigLock - still results in individual features being sequentially-bottlenecked. And when one of those features is "scheduling"...

What it means in practice is that it's not much better than BigLock if you're doing anything that can schedule (and practically all interaction with Zephyr can schedule...)

You can do a whole lot better with "sea of spinlocks + atomics for base scheduling"... at the expense of a whole lot of complexity and nondeterminism. Most RTOSes I've seen don't even attempt this approach - which makes sense in a number of ways - but also means that they tend to be unsuitable for some usecases. The difference between 200 cycles to context-switch when you have a budget of 2000 cycles / core / op and 2000 cycles to context-switch when you have a budget of 2000 cycles / core / op is the difference between "you can carefully use the RTOS" and "you can't use the RTOS", in a lot of ways.

I may end up having to reject Zephyr here, which is rather unfortunate.

I can make a ticket for context-switch performance now to start the discussion, if you'd like. Just know that it's based on a hacked-together branch with MrVan's AArch64 SMP changes, your abort rewrite, some "trivial" local optimizations (notably: changing spinlock release to a sequentially-consistent store instead of a CAS non-trivially improves the performance of the spinlock in contended situations), and some proprietary driver code (nothing that should directly affect this, however)...

Reply:
@andyross - I created #32686 with one of the contributing factors. The current spinlock has some rather terrible behavior under heavy contention (such as all cores hammering at sched_lock).

Reply:
@andyross - I created #32708 with another one of the contributing factors. The current implementation of `k_current_get` is an unfortunate combination of "slow" and "called a lot".